### PR TITLE
Update the docs for 0.6.5 and announce its release.

### PR DESCRIPTION
**Pull request by sjrd**
The announcement is scheduled for Monday 31st morning CEST, assuming everything goes well with the release process.

### CODE REVIEW

**Comment by sjrd** [comment]
Review by everyone ;)
Pinging the usual suspects: @nicolasstucki @gzm0 @lihaoyi 
as well as people who have expressed manifest interest/have already tried Scala.js-defined JS classes: @jducoeur @mysticfall @chandu0101 @russpowers (feel free to Unsubscribe on the right if you don't want to follow this)

**Comment by lihaoyi** [comment]
Looks good to me. So many new things =D I guess after this we'll go and annotate all of scala-js-dom and other facade libraries with `@js.native`?

The visual asymmetry between `@js.native` and `@ScalaJSDefined` bugs me somewhat, but I guess it's no big deal.

**Comment by sjrd** [comment]
> I guess after this we'll go and annotate all of scala-js-dom and other facade libraries with `@js.native`?

Yes ^^

> The visual asymmetry between `@js.native` and `@ScalaJSDefined` bugs me somewhat, but I guess it's no big deal.

It is not nice, but remember that `@ScalaJSDefined` will eventually completely disappear.

**Comment by jducoeur** [comment]
LGTM, and looking forward to it -- should seriously improve facade development...

**Comment by russpowers** [comment]
Looks good, thanks for this really useful release!

**Comment by chandu0101** [comment]
:+1: 

>thanks for this really useful release!
>should seriously improve facade development...

+100 :) 

>Anonymous classes and object literals

I missed this before , a handy one definitely  . Out of curiosity we must implement all fields right ? ,not a replacer for companion objects/builder .. ?

**Comment by chandu0101** [comment]
@sjrd  https://github.com/scala-js/scala-js/pull/1883 is missing

**Comment by sjrd** [comment]
> Out of curiosity we must implement all fields right ? ,not a replacer for companion objects/builder .. ? 

Yes, since Scala.js-defined JS traits cannot have concrete members, the Scala type system will basically force you to implement all advertised members. So for typical options object, that means you have to implement default values as well.

Note that you could provide an `AbstractOptions` inheriting from your `Options`, with concrete implementations for optional values (with `= js.undefined`). Then you can extend that one instead in your object literal, and then you don't have to provide all values anymore :-)

> scala-js/scala-js#1883 is missing

Indeed, thanks. And I also forgot scala-js/scala-js#1818

**Comment by sjrd** [comment]
Artifacts published on Maven Central. ScalaDocs and CLI downloads up on the webserver.

**Comment by ngbinh** [comment]
:dancers: :rose:

**Comment by sjrd** [comment]
Updated with references to `js.constructorOf[C]` (scala-js/scala-js#1883) and the bug scala-js/scala-js#1818.

File: _posts/news/2015-08-31-announcing-scalajs-0.6.5.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+tags: [releases]
+---
+{% include JB/setup %}
+
+We are thrilled to announce the release of Scala.js 0.6.5!
+
+This is probably *the most important release since Scala.js 0.5.0*!
+It brings significant advances to interoperability with JavaScript:
+
+* You can now implement in Scala.js subclasses of JavaScript classes and traits!
```
Comment (gzm0) [inline _posts/news/2015-08-31-announcing-scalajs-0.6.5.md]: Consider putting "in Scala.js" at the end of the phrase.

File: _posts/news/2015-08-31-announcing-scalajs-0.6.5.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+* Using the same semantics, you can now define object literals with `new js.Object { val x = 5 }`
+* There is a new unboxed pseudo-union type `A | B` to type more accurately your JavaScript facades
```
Comment (gzm0) [inline _posts/news/2015-08-31-announcing-scalajs-0.6.5.md]: to more accurately type

File: _posts/news/2015-08-31-announcing-scalajs-0.6.5.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+Addressing these warnings is important to make your source code forward compatible with the next major version of Scala.js.
+An unannotated declaration extending `js.Any` will by default be Scala.js-defined (see next section) in the next version.
+
+## Improvements
+
+### Scala.js-defined JS classes, objects, and traits
+
+Scala.js 0.6.5 introduces a major language improvement: the ability to define, in Scala.js, a subclass of a native JavaScript class (or implementing a JavaScript trait/interface).
+We call such classes *Scala.js-defined JS classes*, because they are effectively JavaScript classes, but written in Scala.js.
+This is opposed to the classes that you typically write, which are Scala classes, not JavaScript classes.
```
Comment (gzm0) [inline _posts/news/2015-08-31-announcing-scalajs-0.6.5.md]: I would not use opposed unless you have a second, dependent phrase. SJS defined classes are not opposed to Scala classes. They complement :)

File: _posts/news/2015-08-31-announcing-scalajs-0.6.5.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+trait Foo extends js.Object {
+  val x: Int     // ok
+  val y: Int = 5 // illegal
+
+  def foo(x: Int): Int         // ok
+  def bar(x: Int): Int = x + 1 // illegal
+}
+{% endhighlight %}
+
+Scala.js-defined JS classes, objects and traits cannot directly extend native JS trait (i.e., non-Scala.js-defined JS traits).
```
Comment (gzm0) [inline _posts/news/2015-08-31-announcing-scalajs-0.6.5.md]: traits

File: _posts/news/2015-08-31-announcing-scalajs-0.6.5.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    x = 5,
+    y = 10
+).asInstanceOf[Position]
+{% endhighlight %}
+
+### Pseudo-union type `A | B`
+
+Many JavaScript libraries have APIs with parameters or values that accept different types of values.
+To be able to accurately type those libraries, Scala.js 0.6.5 features an unboxed, facade-friendly pseudo-union type [`A | B`]({{ site.production_url }}/api/scalajs-library/0.6.5/#scala.scalajs.js.$bar).
+Here are a couple examples of what it can do:
```
Comment (gzm0) [inline _posts/news/2015-08-31-announcing-scalajs-0.6.5.md]: a couple of?

File: _posts/news/2015-08-31-announcing-scalajs-0.6.5.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+  @JSExport val x: Int = 4
+  @JSExport def foo(x: Int): Int = x + 1
+}
+
+SomeLibrary.doSomething(new InterfaceImpl().asInstanceOf[SomeInterface])
+{% endhighlight %}
+
+If you mess up your exports, you will have trouble at run-time.
+
+In 0.6.5, you can write instead:
```
Comment (gzm0) [inline _posts/news/2015-08-31-announcing-scalajs-0.6.5.md]: Consider: "you can write the following instead"

File: _posts/news/2015-08-31-announcing-scalajs-0.6.5.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+{% endhighlight %}
+
+Unlike `x.asInstanceOf[T]`, the `js.use(x).as[T]` idiom *statically typechecks* that you have all the exports required to comply to the JavaScript interface.
+
+### Java library additions
+
+* The complete set of `Character.isXYZ` methods (to test Unicode properties of characters)
+* `java.lang.Math.rint(Double)`
+* `java.util.concurrent.ThreadLocalRandom`
+* `java.util.TreeSet` and `java.util.NavigableView`
```
Comment (gzm0) [inline _posts/news/2015-08-31-announcing-scalajs-0.6.5.md]: NavigableSet. NavigableView is non-standard JDK and implementation private.

File: doc/sjs-defined-js-classes.md [hunk 8]
```diff
@@ -0,0 +1,278 @@
+---
+layout: page
+title: Scala.js-defined JS classes
+---
+{% include JB/setup %}
+
+As explained in detail in the [guide to write facade types](./calling-javascript.html), classes, traits and objects inheriting from `js.Any` are native by default.
+To implement in Scala.js a JavaScript class, it should be annotated with `@ScalaJSDefined`:
```
Comment (gzm0) [inline doc/sjs-defined-js-classes.md]: "To implement a JavaScript class in Scala.js"

**Comment by gzm0** [comment]
That's all. The SJS defined classes section is a bit minimalistic in some spots, but it's probably better to leave it until someone asks anyways :)

**Comment by sjrd** [comment]
OK I will make the changes before merging in 2/3 hours.

**Comment by sjrd** [comment]
Updated.

**Pull request merged**